pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -1342,11 +1436,38 @@ func (ctrl *PersistentVolumeController) provisionClaim(claim *v1.PersistentVolum
 	}
 	klog.V(4).Infof("provisionClaim[%s]: started", claimToClaimKey(claim))
 	opName := fmt.Sprintf("provision-%s[%s]", claimToClaimKey(claim), string(claim.UID))
-	startTime := time.Now()
+	plugin, storageClass, err := ctrl.findProvisionablePlugin(claim)
+	if err != nil {
+		ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, events.ProvisioningFailed, err.Error())
+		klog.Errorf("error finding provisioning plugin for claim %s: %v", claimToClaimKey(claim), err)
+		// failed to find the requested provisioning plugin, directly return err for now.
+		// controller will retry the provisioning in every syncUnboundClaim() call
+		// retain the original behavior of returning nil from provisionClaim call
+		return nil
+	}
 	ctrl.scheduleOperation(opName, func() error {
-		pluginName, err := ctrl.provisionClaimOperation(claim)
-		timeTaken := time.Since(startTime).Seconds()
-		metrics.RecordVolumeOperationMetric(pluginName, "provision", timeTaken, err)
+		// create an opTimestamps without assigning provisionerName/pluginName to get the current timestamp
+		opTimestamps := newOperationTimestamp("")
+		var provisionerName string
+		var err error
+		if plugin == nil || plugin.IsMigratedToCSI() {
+			provisionerName, err = ctrl.provisionClaimOperationCSI(claim, plugin, storageClass)
+		} else {
+			provisionerName, err = ctrl.provisionClaimOperation(claim, plugin, storageClass)
+		}
+		// adding an entry into operationTimestamps for metrics reporting if there exists no
+		// such record for current claim regardless whether the operation returned error or not
+		// add an entry to mark the provisioning request start time
+		opTimestamps.provisionerName = provisionerName
+		claimKey := claimToClaimKey(claim)
+		provisionStartTsKey := ctrl.createOperationTimestampCacheKey("provision", claimKey)
+		if _, exists := ctrl.operationTimestamps.Get(provisionStartTsKey); !exists {
```

Comment:
Can we have cache wrappers that can handle exists entries already existing or not?

Reply:
Maybe it's worth putting these cache related functions into a separate file.

Reply:
> Can we have cache wrappers that can handle exists entries already existing or not?

make sense, will do


> Maybe it's worth putting these cache related functions into a separate file.
We can certainly do that. The only concern is the lifecycle of a cache entry is managed by the controller. Having a separate file only holds the definition, and complicates the initialization of controller. Thus I chose to put into the controller directly as it can be considered as part of it